src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -464,9 +490,102 @@ InvokeHostFunctionOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
         }
     }
 
+    // Apply expiration bumps.
+    for (auto const& bump : out.expiration_bumps)
+    {
+        LedgerKey lk;
+        xdr::xdr_from_opaque(bump.ledger_key.data, lk);
+        releaseAssertOrThrow(isSorobanEntry(lk));
+        auto entryIt = entryRentChanges.find(lk);
+        releaseAssertOrThrow(entryIt != entryRentChanges.end());
+        entryIt->second.newExpirationLedger =
+            std::max(bump.min_expiration, entryIt->second.newExpirationLedger);
+    }
+
+    uint32_t maxExpirationLedger =
+        ledgerSeq + sorobanConfig.stateExpirationSettings().maxEntryExpiration -
+        1;
+    uint32_t minPersistentExpirationLedger =
+        ledgerSeq +
+        sorobanConfig.stateExpirationSettings().minPersistentEntryExpiration -
+        1;
+    uint32_t minTempExpirationLedger =
+        ledgerSeq +
+        sorobanConfig.stateExpirationSettings().minTempEntryExpiration - 1;
+
+    rust::Vec<CxxLedgerEntryRentChange> rustEntryRentChanges;
+    rustEntryRentChanges.reserve(remainingRWKeys.size() +
+                                 resources.footprint.readOnly.size());
+    // Perform actual expiration bumps and compute fee computation inputs.
+    for (auto& [lk, entryChange] : entryRentChanges)
+    {
+        // Skip deleted RW entries.
+        if (!entryChange.readOnly &&
+            remainingRWKeys.find(lk) == remainingRWKeys.end())
+        {
+            continue;
+        }
+        bool isTemporary = isTemporaryEntry(lk);
+        // Enforce minimum expiration for the new entries.
```

Comment:
The minimum is already set by the host, so you could just set `entryChange.newExpirationLedger` to the expiration ledger on the entry.

Reply:
Do we actually need to set it on the host side? I feel like it would be safer if we just had one place where all the validation is performed.

Reply:
Well we either have to set it to the min or 0 on creation, and we thought min made more sense.

Reply:
Actually the host logic is currently wrong (off-by-one error, as usual).

I think we should have just one place that handles all the bump logic. If that's host, then it should do all the deduplication, min/max boundary enforcement and autobump logic, such that the final bump vector can be simply applied to the respective entries. Otherwise, all this work should be performed by core. Doing this on the host side is probably nicer for the sake of preflight/CLI etc., so maybe we should move everything there? I would still do that as a followup though.

Reply:
Yeah we can do that as a followup.